### PR TITLE
crush: document tunables and rule step set_

### PR DESCRIPTION
**Pull request by dachary**

### CODE REVIEW

**Comment by dachary** [comment]
@liewegas after exploring all tunables & the rule steps that can be used to set them, I think only the total_retry tunable + the set_choose_tries step should be documented as useful for the user to modify. The others are for backward compatiblity.

If that looks right to you, I'll turn this into a proper documentation change based on my notes in the description above.

**Comment by dachary** [comment]
@liewegas I got it wrong, please don't review just yet. chooseleaf_tries & choose_tries are both in use, I got confused by the fact that chooseleaf_tries re-uses chooseleaf_descend_once.

**Comment by dachary** [comment]
Here is the updated explanation for the total_retry tunable + the set_choose_tries step + the set_chooseleaf_tries step:

The CHOOSE_FIRSTN and CHOOSE_INDEP rule step look for buckets of a given
type, randomly selecting them. If they are unlucky and find the same
bucket twice, they will try N+1 times (N being the value of the
choose_total_tries tunable). If there is a previous SET_CHOOSE_TRIES
step in the same rule, it will try O times instead (O being the value of
the argument of the SET_CHOOSE_TRIES step).

Note: the choose_total_tries tunable is the number of retry, not the
number of tries. The number of tries is the number of retry + 1. The
SET_CHOOSE_TRIES rule step sets the number of tries and does not need
the + 1. This confusing difference is inherited from an off-by-one bug
from years ago.

The CHOOSELEAF_FIRSTN and CHOOSELEAF_INDEP rule step do the same as
CHOOSE_FIRSTN and CHOOSE_INDEP but also recursively explore each bucket
found, looking for a single device. The same device may be found in two
different buckets because the crush map is not a strict hierarchy, it is
a DAG. When such a collision happens, they will try again. The number of
times they try to find a non colliding device is:

- If FIRSTN and there is no previous SET_CHOOSELEAF_TRIES rule step: try
  N + 1 times (N being the value of the choose_total_tries tunable)

- If FIRSTN and there is a previous SET_CHOOSELEAF_TRIES rule step: try
  P times (P being the value of the argument of the SET_CHOOSELEAF_TRIES
  rule step)

- If INDEP and there is no previous SET_CHOOSELEAF_TRIES rule step: try
  1 time.

- If INDEP and there is a previous SET_CHOOSELEAF_TRIES rule step: try
  P times (P being the value of the argument of the SET_CHOOSELEAF_TRIES
  rule step)

**Comment by liewegas** [comment]
Weird.. the INDEP not respecting set_chooseleaf_tries must have been a bug?

**Comment by dachary** [comment]
@liewegas it's definitely an inconsistency that could lead to unexpected bad mappings. I guess it never led to any complaint because collisions can only happen if the same device shows in two different buckets, which nobody does.

**Comment by dachary** [comment]
@liewegas it is ready for your review when you have time :-)

File: src/crush/builder.h [hunk 1]
```diff
@@ -76,24 +76,64 @@ extern struct crush_rule *crush_make_rule(int len, int ruleset, int type, int mi
...
  * crush_do_rule(), i.e. __arg1__ is __max_result__ minus the number of
  * items already in the result.
  *
+ * - __CRUSH_RULE_SET_CHOOSE_TRIES__ and __CRUSH_RULE_SET_CHOOSELEAF_TRIES__
+ * 
+ *   The CHOOSE_FIRSTN and CHOOSE_INDEP rule step look for buckets of
+ *   a given type, randomly selecting them. If they are unlucky and
+ *   find the same bucket twice, they will try N+1 times (N being the
+ *   value of the choose_total_tries tunable). If there is a previous
+ *   SET_CHOOSE_TRIES step in the same rule, it will try O times
```
Comment (liewegas) [inline src/crush/builder.h]: choose a letter besides O so that it doesn't look like 0 ?

File: src/crush/crush.h [hunk 2]
```diff
@@ -315,28 +315,46 @@ struct crush_map {
...
+         * documentation for crush_rule_set_step() for more
+         * information
+         */
 	__u32 choose_total_tries;
-	/*! attempt chooseleaf inner descent once for firstn mode; on
-	 *! reject retry outer descent.  Note that this does *not*
-	 *! apply to a collision: in that case we will retry as we used
-	 *! to. */
+        /* obsoleted in firefly */
 	__u32 chooseleaf_descend_once;
```
Comment (liewegas) [inline src/crush/crush.h]: I don't understand what the comment means.  The tunables are all here to preserve legacy behavior, so they don't really become obsolete.  And we set it to true now..

File: src/crush/crush.h [hunk 3]
```diff
@@ -315,28 +315,46 @@ struct crush_map {
...
-	/*! if true, it makes chooseleaf firstn to return stable results (if
-	 *! no local retry) so that data migrations would be optimal when some
-	 *! device fails. */
+	/*! Backward compatibility tunable. It is an improvement that
+         *  avoids unnecessary mapping changes, implemented at
+         *  https://github.com/ceph/ceph/pull/6572 and explained in
+         *  this post: "chooseleaf may cause some unnecessary pg
+         *  migrations" in October 2015
+         *  https://www.mail-archive.com/ceph-devel@vger.kernel.org/msg26075.html
+         *  It should always bet set to 1 except for backward compatibility.
```
Comment (liewegas) [inline src/crush/crush.h]: s/bet/be/

File: src/crush/builder.h (hunk 1 above)
Comment (dachary) [inline src/crush/builder.h]: done

File: src/crush/crush.h (hunk 3 above)
Comment (dachary) [inline src/crush/crush.h]: done

File: src/crush/crush.h (hunk 2 above)
Comment (dachary) [inline src/crush/crush.h]: It was a leftover of a misunderstanding on my part, thanks for catching it.

**Comment by dachary** [comment]
updated & repushed

**Comment by dachary** [comment]
@liewegas is there anything else you'd like me to change ?

**Comment by liewegas** [comment]
the s/bet/be/ typo is still there, but otherwise it's good!

**Comment by dachary** [comment]
> the s/bet/be/ typo is still there, but otherwise it's good!

Fixed. I copy pasted the typo 5 times but now I hopefully got all of them :-P

**Pull request merged**